HDDS-7513. [Snapshot] Support list snapshots using fs -ls

xBis7 opened the pull request:
## What changes were proposed in this pull request?

This PR adds support for listing all snapshots under a  bucket by using the `fs -ls` command with providing the `.snapshot` indicator after the bucket name.

```
bash-4.2$ ozone sh volume create /vol1 
bash-4.2$ ozone sh bucket create /vol1/bucket1
bash-4.2$ ozone sh key put /vol1/bucket1/key /etc/hosts
bash-4.2$ ozone sh snapshot create /vol1/bucket1 snap1
bash-4.2$ ozone fs -ls /vol1/bucket1/.snapshot
Found 1 items
drwxrwxrwx   - hadoop hadoop          0 2022-12-28 14:07 /vol1/bucket1/.snapshot/snap1
```

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7513

## How was this patch tested?

For testing this patch, a new test was added under `integration-test/TestOzoneFsSnapshot.java` and a new robot acceptance test was created.

GeorgeJahad commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OFSPath.java:
```diff
@@ -263,6 +264,23 @@ public boolean isBucket() {
+   * only contains the snapshot indicator, then return true.
+   * e.g. /vol/bucket/.snapshot is a snapshot prefix.
+   */
+  public boolean isSnapshotPrefix() {
```
we have used the term "prefix" in so many different contexts it is a bit confusing.  What do you think about calling this method "isSnapshotIndicator"  to be consistent with the OM_SNAPSHOT_INDICATOR constant used below?

GeorgeJahad commented on hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java:
```diff
@@ -603,12 +605,24 @@ public FileStatusAdapter getFileStatus(String path, URI uri,
       return getFileStatusAdapterForRoot(uri);
     }
+
+    OzoneVolume volume = objectStore.getVolume(ofsPath.getVolumeName());
```
I'm guessing this line was moved here out of the if statement below because it is needed elsewhere.  That is often a good idea, but I think I remember people saying they were concerned about the performance of getFileStatus(), so I would leave it where it was and invoke it again from within the
```
if (ofsPath.isSnapshotPrefix()) {
```
block.

GeorgeJahad commented on hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java:
```diff
@@ -603,12 +605,24 @@ public FileStatusAdapter getFileStatus(String path, URI uri,
       OzoneBucket bucket = getBucket(ofsPath, false);
+
+      if (ofsPath.isSnapshotPrefix()) {
+        UserGroupInformation ugi =
+            UserGroupInformation.createRemoteUser(volume.getOwner());
+        String owner = ugi.getShortUserName();
+        String group = getGroupName(ugi);
```
I would move these lines into getFileStatusAdapterWithSnapshotPrefix()

GeorgeJahad commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFsSnapshot.java:
```diff
@@ -61,59 +77,69 @@ public static void initClass() throws Exception {
+    res = ToolRunner.run(shell,
+        new String[]{"-ls", BUCKET_ONE_WITH_SNAPSHOT_PREFIX_PATH});
+    // Asserts that -ls request succeeded
+    Assertions.assertEquals(0, res);
```
Can the results be checked with "GenericTestUtils.SystemOutCapturer"?

GeorgeJahad commented on hadoop-ozone/dist/src/main/smoketest/snapshot/fs-snapshot.robot:
```diff
@@ -0,0 +1,78 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
```
I appreciate you adding a smoke test.  I should have done it in my PR (but was too lazy).

Since "no good deed goes unpunished", I'm going to ask you to do a bit more.

Please do an "fs -ls" on "$snapshot1"  and on "$snapshot1/key1" and confirm that they work correctly as well.

xBis7 commented:
@smengcl I rebased this branch as well and resolved the conflicts. In the PR's new robot test, I added a check for `ozone fs -ls` for snapshot keys. I noticed that after rebasing this branch, that command hangs up and doesn't produce an output. I suspect it has to do with the merging changes in `HDDS-6517-Snapshot`. Nevertheless, the robot test will timeout and fail in the workflow.

Are there more merging changes coming in `HDDS-6517-Snapshot` that might fix this? Should I remove the key tests from the robot file?

GeorgeJahad commented on hadoop-ozone/dist/src/main/smoketest/snapshot/fs-snapshot.robot:
```diff
@@ -0,0 +1,95 @@
+                    Set Suite Variable     ${DIR_ONE}    dir-${random}
+    ${random} =     Generate Random String  5  [LOWER]
+                    Set Suite Variable     ${KEY_TWO}    key-${random}
+    ${result} =     Execute             ozone sh key put /${VOLUME}/${BUCKET}/${DIR_ONE}/${KEY_TWO} HISTORY.md
```
This key doesn't seem to get used.

GeorgeJahad commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFsSnapshot.java:
```diff
@@ -61,59 +78,78 @@ public static void initClass() throws Exception {
+      // Assert that output contains snap2
+      Assertions.assertTrue(capture.getOutput()
+          .contains(BUCKET_ONE_WITH_SNAPSHOT_INDICATOR_PATH
+              + OM_KEY_PREFIX + SNAPSHOT_TWO));
```
Check for KEY_TWO as well

GeorgeJahad commented:
> This could have something to do with the rebase due to new OM#getKeyInfo() method

Yes that is the problem.  I hope to have a solution in the next day or two.  Thanks for pointing it out @smengcl @xBis7 @aswinshakil

xBis7 commented:
@GeorgeJahad Thanks for the review! I've addressed your comments. I will revisit this PR to fix any errors related to my changes after rebasing `HDDS-6517-Snapshot` has finished.

smengcl commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OFSPath.java:
```diff
@@ -263,6 +264,23 @@ public boolean isBucket() {
+   * e.g. /vol/bucket/.snapshot is a snapshot indicator.
+   */
+  public boolean isSnapshotIndicator() {
+    if (keyName.contains(OM_SNAPSHOT_INDICATOR)) {
```
I feel like we should check for key prefix (`OM_KEY_PREFIX + OM_SNAPSHOT_INDICATOR` ?) rather than `OM_SNAPSHOT_INDICATOR`.

Because we are not preventing users from creating a regular key with `.snapshot` in its name?

xBis7 replied:
@smengcl I get what you are saying, you are right! Key name might be `sn.snapshot` and this method would still accept it. I'll fix it!

hemantk-12 replied:
Just curious if we should prevent user from creating key with `.snapshot` in it because it is kind of reserved keyword for Ozone.

GeorgeJahad replied:
there is a ticket for that here: https://issues.apache.org/jira/browse/HDDS-7168

hemantk-12 replied:
Thanks @GeorgeJahad

hemantk-12 commented on hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java:
```diff
@@ -643,6 +645,13 @@ public FileStatusAdapter getFileStatus(String path, URI uri,
     try {
       OzoneBucket bucket = getBucket(ofsPath, false);
+
+      if (ofsPath.isSnapshotIndicator()) {
```
It would be easier to read if written like below and exception is handle in individual function.
```
    if (ofsPath.isRoot()) {
      return getFileStatusAdapterForRoot(uri);
    } else if (ofsPath.isVolume()) {
      OzoneVolume volume = objectStore.getVolume(ofsPath.getVolumeName());
      return getFileStatusAdapterForVolume(volume, uri);
    } else if (ofsPath.isSnapshotIndicator()) {
      OzoneVolume volume = objectStore.getVolume(ofsPath.getVolumeName());
      return getFileStatusAdapterWithSnapshotIndicator(
          volume, bucket, uri);
    } else {
      OzoneFileStatus status = bucket.getFileStatus(key);
      return toFileStatusAdapter(status, userName, uri, qualifiedPath,
          ofsPath.getNonKeyPath());
    }
```

xBis7 replied:
I think it's worth keeping the try catch, in order to have a more specific error message in case of failure. It's gonna look like so 

```suggestion
      if (ofsPath.isRoot()) {
        ...
      } else if (ofsPath.isVolume()) {
        ...
      } else {
        try {
          if (ofsPath.isSnapshotIndicator()) {
            ...
          } else {
            ...
          }
        } catch () {
          ...
        } 
      }
```

hemantk-12 replied:
Don't have strong point to move try-catch to individual function. One thing it will improve is nesting. Right now we have too much nesting.

xBis7 replied:
@hemantk-12 I get that you want to get rid of nesting. I don't understand what you mean to do with the try-catch. Do you want me to wrap all the `if - else if` block with the try-catch or just wrap whatever line might throw an exception? Or do you want me to remove the try-catch altogether?

xBis7 replied:
ooh, now I get it. You are saying to move 
```
        try {
          if (ofsPath.isSnapshotIndicator()) {
            ...
          } else {
            ...
          }
        } catch () {
          ...
        } 
```
to a new function.

hemantk-12 commented on hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java:
```diff
@@ -782,6 +791,32 @@ private List<FileStatusAdapter> listStatusVolume(String volumeStr,
   }
 
+  /**
+   * Helper for OFS listStatus
```
```suggestion
   * Helper for OFS listStatus on a bucket to get all snapshots.
 ```
 
 Could be a one line.

hemantk-12 commented on hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java:
```diff
@@ -782,6 +791,32 @@ private List<FileStatusAdapter> listStatusVolume(String volumeStr,
+    String group = getGroupName(ugi);
+
+    OzoneBucket ozoneBucket = getBucket(volumeName, bucketName, false);
+    List<FileStatusAdapter> fileStatusAdapterList = new ArrayList<>();
+    for (OzoneSnapshot ozoneSnapshot : snapshotList) {
+      fileStatusAdapterList.add(
+          getFileStatusAdapterForBucketSnapshot(
+              ozoneBucket, ozoneSnapshot, uri, owner, group));
+    }
+
+    return fileStatusAdapterList;
```
```suggestion
    return snapshots.stream()
        .map(ozoneSnapshot -> getFileStatusAdapterForBucketSnapshot(
            ozoneBucket, ozoneSnapshot, uri, owner, group))
        .collect(Collectors.toList());
```
nit: you can use lambdas.

hemantk-12 reviewed (commented):
Thanks for the patch @xBis7

xBis7 commented on hadoop-ozone/dist/src/main/smoketest/snapshot/fs-snapshot.robot:
```diff
@@ -0,0 +1,96 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
```
We are testing different operations but we could have a common setup robot file and reuse that. Isn't it cleaner to have different tests for `sh` commands and `fs` commands?

#4171 is simpler and more straight forward. We could merge it and then refactor it in this PR and create a common setup robot file for setting up volumes, buckets, snapshots and reuse these keywords in both acceptance tests.

hemantk-12 replied:
Sounds good to me.

xBis7 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFsSnapshot.java:
```diff
@@ -24,33 +25,38 @@
+  private static String bucketPath;
+  private static String bucketWithSnapshotIndicatorPath;
+
+  @BeforeAll
```
I could group most of them into successful and failed snapshot creations.

xBis7 commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OFSPath.java:
```diff
@@ -263,6 +264,23 @@ public boolean isBucket() {
+   * only contains the snapshot indicator, then return true.
+   * e.g. /vol/bucket/.snapshot is a snapshot indicator.
+   */
+  public boolean isSnapshotIndicator() {
```
Is the snapshot indicator pointing to an actual directory or only used for referring to a snapshot? If it's not, this change is confusing.

hemantk-12 replied:
May be you can use something more generic like `isShanshotAddress` or more relevant. But `isSnapshotIndicator()` is bit a confusing to me.

xBis7 replied:
How about `isSnapshotPath()`?

hemantk-12 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFsSnapshot.java:
```diff
@@ -159,63 +141,53 @@ public void testCreateSnapshotDuplicateName() throws Exception {
+      "snap156808943643007724443266605711479126926050896107709081166294"})
+  public void testCreateSnapshotFailure(String snapshotName)
+      throws Exception {
+    if (snapshotName.equals("invalidURI")) {
```
nit: you could have pass `bucketPath` as well instead of tweaking it based on `snapshotName`.

xBis7 replied:
I tried it both ways and neither seem a clean approach, so I restored the original method that checks that the creation fails due to the volume-bucket URI. After all, this has to do with the path and not the snapshot name so it makes sense to have it in a seperate method.

xBis7 commented:
@hemantk-12 @smengcl I've refactored the acceptance test from #4171 and created a common setup file for snapshot acceptance tests as discussed [here](https://github.com/apache/ozone/pull/4134#discussion_r1085311631). Can you please take a look?

smengcl commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/IncrementalContainerReportHandler.java:
```diff
@@ -88,8 +88,10 @@ public void onMessage(final IncrementalContainerReportFromDatanode report,
             id = container.containerID();
           } finally {
-            if (!replicaProto.getState().equals(
+            if (replicaProto.getState().equals(
                 ContainerReplicaProto.State.DELETED)) {
+              nodeManager.removeContainer(dd, id);
+            } else {
```
This looks like a GitHub PR diff display bug, which makes it look like it is brought in by the merge commit https://github.com/apache/ozone/pull/4134/commits/e7325c3b6d737d90f4dd647bcfd2ce567a1cab43

while the `master` branch already has the change: https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/IncrementalContainerReportHandler.java#L90-L97

It should be safe to ignore this.

xBis7 replied:
@smengcl Fixed it!

hemantk-12 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFsSnapshot.java:
```diff
@@ -65,259 +77,229 @@ public static void initClass() throws Exception {
-    String bucket = "buc-" + RandomStringUtils.randomNumeric(5);
-    String key = "key-" + RandomStringUtils.randomNumeric(5);
-    String snapshotName = "snapa?b";
+  /**
```
How case 3 is getting test? You are not passing or setting `bucketPath` it to invalid bucket? I'll suggest something like following:

```suggestion
  private static Stream<Arguments> createSnapshotFailureScenarios() {
    return Stream.of(
        Arguments.of("1st case: snapshot name contains invalid char",
            validBucketPath,
            "snapa?b"),
        Arguments.of("2nd case: snapshot name consists only of numbers",
            validBucketPath,
            "1234"),
        Arguments.of("3rd case: bucket path is invalid",
            invalidBucketPath,
            "validSnapshotName12"),
        Arguments.of("4th case: snapshot name length is more than 64 chars",
            validBucketPath,
            "snap156808943643007724443266605711479126926050896107709081166294")
    );
  }

  @ParameterizedTest(name = "{0}")
  @MethodSource("createSnapshotFailureScenarios")
  public void testCreateSnapshotFailure(String description,
                                        String bucketPath,
                                        String snapshotName)
      throws Exception {

    int res = ToolRunner.run(shell,
        new String[]{"-createSnapshot", bucketPath, snapshotName});
    Assertions.assertEquals(1, res);
  }
```

Also is there a way to validate the response from OFS. Because we are just validating that request fails but not verify the reason. We should verify that to make sure that reason is correct.

xBis7 replied:
> Also is there a way to validate the response from OFS. Because we are just validating that request fails but not verify the reason. We should verify that to make sure that reason is correct.

We can use `GenericTestUtils.SystemOutCapturer`, I'll pass the expected message in the parameters and fix it.

hemantk-12 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFsSnapshot.java:
```diff
@@ -65,259 +77,229 @@ public static void initClass() throws Exception {
-      // Asserts that create request failed since invalid name passed
-      assertEquals(1, res);
+  @Test
+  public void testCreateSnapshotInvalidURI() throws Exception {
```
This won't be needed if you create tests as suggested in previews comment.

hemantk-12 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFsSnapshot.java:
```diff
@@ -65,259 +77,229 @@ public static void initClass() throws Exception {
-    String testVolBucket = OM_KEY_PREFIX + volume + OM_KEY_PREFIX + bucket;
-    String testKey = testVolBucket + OM_KEY_PREFIX + key;
+  @Test
+  public void testCreateSnapshotParameterMissing() throws Exception {
```
What if we pass empty string for URI? Would it be wrong bucket path or bucket path is missing.

Just thinking if all the failure scenarios can be group to single parameterized test.

xBis7 replied:
I can't group it with the rest because it expects `-1` as the result integer instead of `1`. It actually tests `createSnapshot` without any parameters, since a snapshot without a name can be created.

hemantk-12 replied:
Optional: you could pass expected response in parameters similar to `expectedMessage`.
e.g.

```
  public void testCreateSnapshotFailure(String description,
                                        String paramBucketPath,
                                        String snapshotName,
                                        String expectedMessage,
                                        int expectedShellResponse)
```

xBis7 replied:
I can fix that!

hemantk-12 replied:
Thanks @xBis7

hemantk-12 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFsSnapshot.java:
```diff
@@ -65,259 +77,229 @@ public static void initClass() throws Exception {
-
-    String testVolBucket = OM_KEY_PREFIX + volume + OM_KEY_PREFIX + bucket;
-    String testKey = testVolBucket + OM_KEY_PREFIX + key;
+  public void testFsLsSnapshot() throws Exception {
```
nit: I think `testFsLsSnapshot` and `testFsLsSnapshotKeys` can be grouped.

hemantk-12 commented on hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java:
```diff
@@ -635,19 +637,36 @@ public FileStatusAdapter getFileStatus(String path, URI uri,
     }
+  }
+
+  /**
```
nit:
```suggestion
  /**
   * Return FileStatusAdapter based on OFSPath is a valid bucket path
   * or valid snapshot path.
   * Throws exception in case of failure.
   */
```

hemantk-12 commented on hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java:
```diff
@@ -635,19 +637,36 @@ public FileStatusAdapter getFileStatus(String path, URI uri,
-          ofsPath.getNonKeyPath());
+      if (ofsPath.isSnapshotPath()) {
+        OzoneVolume volume = objectStore.getVolume(ofsPath.getVolumeName());
+        return getFileStatusAdapterWithSnapshotIndicator(
```
nit: Should we use`toFileStatusAdapterWithSnapshotIndicator` just to be consistent with `toFileStatusAdapter`?

xBis7 replied:
@hemantk-12 Skipping this part, will lead to failure as it will consider the snapshot indicator as a key.

hemantk-12 replied:
No problem.

hemantk-12 commented on hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java:
```diff
@@ -1123,6 +1169,75 @@ private static FileStatusAdapter getFileStatusAdapterForBucket(
+   * @param uri Full URI to OFS root.
+   * @return FileStatusAdapter for a snapshot indicator.
+   */
+  private static FileStatusAdapter getFileStatusAdapterWithSnapshotIndicator(
```
What's difference between `getFileStatusAdapterWithSnapshotIndicator` and `getFileStatusAdapterForBucketSnapshot`? Also `getFileStatusAdapterWithSnapshotIndicator`'s description says `Generate a FileStatusAdapter for a bucket` but it is for a snapshot bucket or at least snapshot path.

xBis7 replied:
Let's say that we have `vol1`, `bucket1` and `snap1`. 

If we are on path `/vol1/bucket1/.snapshot/snap1` and we need to get the `FileStatusAdapter` for `snap1`, we will call `getFileStatusAdapterForBucketSnapshot`. 

If we are on path `/vol1/bucket1/.snapshot`, then we need to get the `FileStatusAdapter` for the snapshot indicator (`.snapshot`) and we will call `getFileStatusAdapterWithSnapshotIndicator`. The snapshot indicator isn't a directory or an entity like a bucket or a snapshot or a key but just a prefix for OM to know that we are referring to a snapshot and it should look it up in the snapshot table. `/vol1/bucket1/.snapshot` isn't an actual path and we can't have a `FileStatusAdapter` for the indicator. Therefore, we create a `FileStatusAdapter` for the bucket and append the snapshot indicator to the path and that's the difference with the `FileStatusAdapter` for a bucket.

hemantk-12 replied:
I see. Thanks for the explanation.

xBis7 commented:
@hemantk-12 Regarding `TestOzoneFsSnapshot`, I couldn't reuse `GenericTestUtils.SystemOutCapturer` more than once and there were many test where we needed to check the shell output, so I wrote a method with similar behavior. Let me know how it looks.

smengcl commented on hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java:
```diff
@@ -784,6 +804,28 @@ private List<FileStatusAdapter> listStatusVolume(String volumeStr,
+
+    OzoneVolume volume = objectStore.getVolume(volumeName);
+    UserGroupInformation ugi =
+        UserGroupInformation.createRemoteUser(volume.getOwner());
```
Should we fill in bucket owner instead?

xBis7 replied:
yes, I'll fix that.

smengcl commented on hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java:
```diff
@@ -784,6 +804,28 @@ private List<FileStatusAdapter> listStatusVolume(String volumeStr,
+  private List<FileStatusAdapter> listStatusBucketSnapshot(
+      String volumeName, String bucketName, URI uri) throws IOException {
+    List<OzoneSnapshot> snapshotList =
+        objectStore.listSnapshot(volumeName, bucketName);
```
I assume this `listSnapshot` by default filters out all snapshot that are not `ACTIVE` ? For example some snapshots are deleted pending reclaimation would still remain in the OM table.

xBis7 replied:
I didn't create `listSnapshot()`, [here](https://github.com/xBis7/ozone/blob/HDDS-7513/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java#L1200-L1206) is the implementation. There are other places or commands using this method. If we add the filter below here, we wouldn't be that consistent but on the other hand we might need `listSnapshot()` to return all the snapshots and not just the `ACTIVE` ones.

smengcl replied:
Agreed. We should do this in another JIRA.

smengcl reviewed (commented):
Looks good overall.

hemantk-12 commented:
> @hemantk-12 Regarding `TestOzoneFsSnapshot`, I couldn't reuse `GenericTestUtils.SystemOutCapturer` more than once and there were many test where we needed to check the shell output, so I wrote a method with similar behavior. Let me know how it looks.

That's bad because in your previous commit, tests were failing because of different reason than expected.

hemantk-12 commented:
Overall lgtm from my side.

hemantk-12 reviewed (approved):
LGTM.

xBis7 commented:
> That's bad because in your previous commit, tests were failing because of different reason than expected.

@hemantk-12 Do you mean [this commit](https://github.com/apache/ozone/pull/4134/commits/8d7fcdf6dcb6bac06c75ad12ba52e187b2c9a31a#diff-03616b28a7587e3f7f89f8a9b3fc268174854e8d32c37ff9c06bb20fda4150c5)?

hemantk-12 commented:
> > That's bad because in your previous commit, tests were failing because of different reason than expected.
> 
> @hemantk-12 Do you mean [this commit](https://github.com/apache/ozone/pull/4134/commits/8d7fcdf6dcb6bac06c75ad12ba52e187b2c9a31a#diff-03616b28a7587e3f7f89f8a9b3fc268174854e8d32c37ff9c06bb20fda4150c5)?

Sorry, I couldn't find the actual commit.
You can leave these tests as it is for now. For future, we should think how we can check the actual failure reason and validate that.

xBis7 commented:
> For future, we should think how we can check the actual failure reason and validate that.

@hemantk-12 In this case, the `GenericTestUtils.SystemOutCapturer` was capturing the output but it wasn't restoring the stream afterwards and `System.setOut()` wasn't getting restored.

For instance, if we were to run 2-3 commands in one test and wanted to check the output for each and every one of them, the capturer would get the output of the first command and it would keep it for every new command. If we tried to close it and then create a new one, it would get stuck closing and the output would be empty for the other commands or tests.

In `execShellCommandAndGetOutput`, we reset the stream [here](https://github.com/apache/ozone/blob/68fca589147b07dc34f36ca46b86f281a45390b0/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFsSnapshot.java#L307-L315), so that `System.setOut()` has an empty `ByteArrayOutputStream` and we can set a new value on the next method call.

prashantpogde commented:
@xBis7 Can you check the CI failure ? it should be ready for merge post clean CI.

xBis7 commented:
@prashantpogde The failure is unrelated to the changes. I've got a green CI on my fork, check the workflow [here](https://github.com/xBis7/ozone/actions/runs/4096060601). Can you please rerun just the failed test? It should pass.

smengcl commented:
CI passed after retriggering. Thanks @xBis7 for the patch. Thanks @GeorgeJahad @hemantk-12 @prashantpogde for reviewing.
